accepted/future-releases/nnbd/feature-specification.md

```diff
@@ -263,6 +268,157 @@ The **flatten** function is modified as follows:
 and for all `R`, if `T <: Future<R>` then `S <: R`; then **flatten**(`T`) = `S`
   - otherwise **flatten**(`T`) = `T`
 
+### The future value type of an asynchronous generator
+
+_We specify a concept which corresponds to the type of object which is
+delivered by an `async` function, that is, the type of object which is used to
+complete the future which is returned from such a function when it suspends
+for the first time._
+
+Let _f_ be an asynchronous non-generator function. We define the notion of the
+**future value type** of _f_ as follows: If, for some `U`, the declared return
+type `T` of _f_ is `Future<U>`, `FutureOr<U>`, `Future<U>?`, or `FutureOr<U>?`,
+the future value type of _f_ is `U`.
```

Comment:
I believe we have a split here:

Normalization is applied in just two specific situations according to the nnbd **spec** (mixed inheritance conflict mitigation and reified `Type` equality).

However, the **implementations** (certainly the CFE, probably the analyzer as well) will apply normalization in several very common situations in addition to the specified situations: The nullability of a type is represented as the value of a field (not as an operator on a type, yielding a wrapped object that represents the type, and a wrapper object that represents the `*` or `?`). One consequence of this representation choice is that it is impossible for a CFE representation of a type to model `T**` for any `T`; when we add the second `*`, it amounts to setting the nullability field to 'legacy' (I think it's called `indefinite`), and setting that field the second time is a normalization, because we can't see that it was done twice. Similarly, if the nullability was 'nullable' (`?`) and it is made 'legacy' it will just remain 'nullable'.

This means that normalization is used very sparingly in the spec, but it is also used very frequently in certain situations. Luckily, those very frequent non-specified normalizations are probably benign. I do not think that the "heavier" normalizations like `FutureOr<dynamic>` --> `dynamic` will ever occur except in the specified situations.

This means that in practice we will never see `Future<int>??` and such.

However, I'll reword the text in such a way that it _will_ work with the spec model (where `T**` can exist). Definitely that would be the safe approach right now.

That said, we probably don't have to insist on aligning the implementations and the spec in this respect, as long as we are aware of the fact that we may predict a type like `T**` based on the spec, and the implementations will insist that it's `T*`. It is going to be tricky to observe the difference, because normalization will never introduce nor eliminate any subtyping relationships (we always have `T <: NORM(T) <: T`).

Reply:
So the spec needs to specify what *flatten*(FutureOr<int?*?>*?*)` is too, and everywhere we say "of the form ..." we need to account or any number of modifiers afterwards?

Reply:
Alternatively, we could have more than one kind of normalization: The "small scale" normalization that just replaces `**` by `*` and so on everywhere and the "large scale" that replaces actual type structure (like `FutureOr<T>` --> `T` for some `T`), which is only applied.

We could also try to argue that the difference is not observable (except for some unspecified cases like `print(something.runtimeType)`). The subtype rules will make most of the distinctions go away, because of the `T <: NORM(T) <: T` property, so we can only see the difference when there is no subtyping step involved, e.g., in declarations of the superinterfaces of a class.

Reply:
The futureValueType function was introduced and rules rewritten to handle non-normalized types.